src/ic-cdk/src/futures.rs

```diff
@@ -1,121 +1,135 @@
-        // The box is never moved out of, but dropped in place.
-        let mut pinned_future = unsafe { Pin::new_unchecked(&mut *boxed_future) };
-        if !super::CLEANUP.load(Ordering::Relaxed)
-            && pinned_future
-                .as_mut()
-                // SAFETY: The pointer attached via `waker` satisfies its own contract.
-                .poll(&mut Context::from_waker(&unsafe { waker::waker(ptr) }))
-                .is_pending()
-        {
-            Box::into_raw(boxed_future_ptr_ptr);
-            Box::into_raw(boxed_future);
+        // SAFETY: The function's contract guarantees that the pointer is an Rc to a WakerState, and that this call takes ownership of the data.
+        let state = unsafe { Rc::from_raw(ptr as *const WakerState) };
+        // Must check CLEANUP *before* previous_trap, as we may be recovering from the following immediate trap.
+        if super::CLEANUP.load(Ordering::Relaxed) {
+            state.previous_trap.set(true);
+        } else if state.previous_trap.get() {
+            crate::trap("Call already trapped");
+        } else {
+            let waker = waker(Rc::clone(&state));
+            let Ok(mut borrow) = state.future.try_borrow_mut() else {
+                // If this is already borrowed, then wake was called from inside poll. There's not a lot we can do about this - we are not
+                // a true scheduler and so cannot immediately schedule another poll, nor can we reentrantly lock the future. So we cancel it.
+                // Don't wake from poll. There's really no reason to do that at all. 
+                return;
+            };
+            let pinned_future = borrow.as_mut();
+            let _ = pinned_future.poll(&mut Context::from_waker(&waker));
         }
     }
 
-    fn wake_by_ref(_: *const ()) {}
-
-    fn drop(_: *const ()) {}
+    /// # Safety
+    ///
+    /// This function should only be called by a [Waker] created by [waker].
+    unsafe fn wake_by_ref(ptr: *const ()) {
+        // SAFETY:
+        // The function's contract guarantees that the pointer is an Rc to a WakerState, and that this call borrows the data.
+        // wake has the same contract, except it takes ownership instead of borrowing. Which just requires incrementing the refcount.
+        unsafe {
+            Rc::increment_strong_count(ptr as *const WakerState);
+            wake(ptr);
+        }
+    }
 
     /// # Safety
     ///
-    /// The pointer must be a unique Box-allocated pointer to a Box-allocated pointer to a `dyn Future<Output=()>`.
-    pub unsafe fn waker(ptr: *const ()) -> Waker {
+    /// This function should only be called by a [Waker] created by [waker].
+    unsafe fn drop(ptr: *const ()) {
+        // SAFETY: The function contract guarantees that the pointer is an Rc to a WakerState, and that this call takes ownership of the data.
+        unsafe {
+            Rc::from_raw(ptr as *const WakerState);
```

Comment:
Would it be clearer to use decrement_strong_count which seems more aligned with the purpose of this line?

Reply:
It would be possible to, but I do not think it would be clearer to; discarding the result of `from_raw` is very recognizable as calling a destructor, as that is how you do it with Box or Vec.